packages/lib/src/interactions/SelectionTool.tsx

```diff
@@ -131,10 +129,30 @@ function SelectionTool(props: Props) {
   );
 
   useEffect(() => {
-    if (selection && isModifierKeyPressed) {
-      onSelectionChangeRef.current?.(selection);
+    if (!prevSelection && selection) {
+      onSelectionStartRef.current?.(selection);
+      return;
+    }
+
+    if (prevSelection && !selection && hasSuccessfullyEndedRef.current) {
```

Comment:
Not sure if this is the effect of a December Friday afternoon but these conditions feel hard to process for me :joy_cat:  .

I would suggest to add some comments but perhaps having the explanation later will make things crystal clear :smile_cat:

Reply:
`prevSelection` holds the selection of the previous render.

- When the selection starts, it's `undefined` and the current `selection` is defined, so we know to call `onSelectionStart`.
- When the selection continues, it's defined, and the current `selection` is also defined, so we know to call `onSelectionChange`.
- When the selection ends, it's defined but the current `selection` is `undefined`, so we know to call `onSelectionEnd`.

`hasSuccessfullyEnded` is just a flag so we don't call `onSelectionEnd`:

- when the user actually cancels the selection with `Escape`; or
- when the call to `shouldInteract` returns `false` in `onPointerUp` (i.e. when the user has released a modifier key before releasing the pointer).

Reply:
Yeah I wonder if making  `hasSuccessfullyEnded` smarter could simplify all this. 

It could be containing the current status of the selection process (e.g. `started`, `inProgress`, `ended`, `invalid`) so that we don't rely on `selection === undefined` that is an ambiguous condition that needs further tests to know the exact state of the selecting process.

Reply:
The ref solution works because it tracks a very specific thing. If I start complexifying it, I run the risk of introducing race conditions (e.g. what if `onPointerMove` is called before the `useEffect` has had a chance to run?) I could store a more complex `selection` state with a `status` key for instance, but in that case, we'd be better off using a full-on state machine, otherwise we're sure to re-introduce tricky state-related bugs. I felt the ref was a simple, bug-free solution for the problem at hand.

Reply:
I understand. But the chain of conditions feel really convoluted to me so I feel it should at least be commented.

Reply:
I've added a bunch of comments, let me know if they make sense.

Reply:
They do. Thanks :+1: